### DISCUSSION TITLE
Authorisation below the page level

### DISCUSSION THREAD

**Discussion by iainhallam** [lowdefy/lowdefy#1428] [Ideas and feature requests]
I'm thinking about Lowdefy to use to build a tool for members of a club to see information related to their groups. Groups shouldn't be able to see each others' info, and members of the club can be in more than one group, and create new ones from time to time, so I'd prefer this to be done in data, rather than as part of the app structure.

Thinking about the basic pages needed, I start out with `/groups` to list the groups you're in, but then when you want to see the content information for that group, I guess with the current scheme you'd need to get to `/groups?group=group-id`. At this point, I don't think I can model the authorisation needed using the per-page protection, unless there's some other way to specify that only members of that group can see its data. Can blocks implement some kind of role or membership checking?

Is there a way to achieve this in Lowdefy?

**Comment by vicam001** [comment]
Hi @iainhallam , 

Very interesting need. I am also starting to find more challenging requirements in my Lowdefy apps when it comes to permissions and access to data. I can recommend you have a look at **Hasura GraphQL Engine** as a way to act as a middleware between your Lowdefy app and the data source (e.g. a Postgres database), this way instead of relaying on the database user used in the connection, you can use the user attributes (e.g. tenant_id, member_of_groups,...) to retrieve just the right data (and/or restrict which operation the user is allowed to perform, such as being able to SELECT and INSERT but not UPDATE and DELETE and similar stuff).

**Comment by SamTolmay** [comment]
Hi @iainhallam 

This can be done in Lowdefy, but does require a few things. I will be discussing how to do things in Lowdefy V4, similar can be done for  V3.

We add attributes to the user from our provider. Specifically, we add the attributes to the user token in Auth0 using a login rule, but this can also be done in other ways, like a Next Auth adapter for example. 

We then use these attributes in requests using the `_user` operator. This is secure becuase the request is evaluated on the Lowdefy server. So for example we would do something like:

```yaml
id: fetch_groups
type: MongoDBFind
connectionId: groups
properties:
  query:
    _id:
      $in:
        _user: group_ids 
```

You could also use Lowdefy custom requests or a different API server if you need something more advanced.
